optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/manager/DefaultSolverManager.java

```diff
@@ -0,0 +1,207 @@
+
+    private DefaultSolverManager(String solverConfigResource, ClassLoader classLoader, ThreadFactory threadFactory) {
+        if (classLoader != null) {
+            solverFactory = SolverFactory.createFromXmlResource(solverConfigResource, classLoader);
+        } else {
+            solverFactory = SolverFactory.createFromXmlResource(solverConfigResource);
+        }
+
+        problemIdToSolverTaskMap = new ConcurrentHashMap<>();
+        int numAvailableProcessors = Runtime.getRuntime().availableProcessors();
+        int activeThreadCount = Math.max(numAvailableProcessors - 2, 1);
+        logger.info("Number of available processors: {}. Active thread count: {}.", numAvailableProcessors, activeThreadCount);
+
+        if (threadFactory != null) {
+            solverFactory.getSolverConfig().setThreadFactoryClass(threadFactory.getClass());
+            solverExecutorService = Executors.newFixedThreadPool(activeThreadCount, threadFactory); // #threads < #cpus implies a FIFO scheduling policy
+            eventHandlerExecutorService = Executors.newSingleThreadExecutor(threadFactory);
+        } else {
+            solverExecutorService = Executors.newFixedThreadPool(numAvailableProcessors - 1);
+            eventHandlerExecutorService = Executors.newSingleThreadExecutor();
+        }
+    }
+
+    @Override
+    public void solve(Object problemId,
+                      Solution_ planningProblem,
+                      Consumer<Solution_> onBestSolutionChangedEvent,
+                      Consumer<Solution_> onSolvingEnded) {
+        solve(problemId, planningProblem, onBestSolutionChangedEvent, onSolvingEnded, null);
+    }
+
+    @Override
+    public void solve(Object problemId,
+                      Solution_ planningProblem,
+                      Consumer<Solution_> onBestSolutionChangedEvent,
+                      Consumer<Solution_> onSolvingEnded,
+                      Consumer<Throwable> onException) {
+        SolverTask<Solution_> newSolverTask;
+        synchronized (this) {
+            if (isProblemSubmitted(problemId)) {
+                throw new IllegalArgumentException("Problem (" + problemId + ") already exists.");
+            }
+            newSolverTask = new SolverTask<>(problemId, solverFactory.buildSolver(), planningProblem);
+            problemIdToSolverTaskMap.put(problemId, newSolverTask);
+            logger.info("A new solver task was created with problemId ({}).", problemId);
+        }
+
+        // TODO implement throttling
+        if (onBestSolutionChangedEvent != null) {
+            newSolverTask.addEventListener(
+                    bestSolutionChangedEvent ->
+                            eventHandlerExecutorService.submit(
+                                    () -> onBestSolutionChangedEvent.accept(bestSolutionChangedEvent.getNewBestSolution())));
+        }
+
+        CompletableFuture<Solution_> solverFuture = CompletableFuture.supplyAsync(newSolverTask::startSolving, solverExecutorService);
```

Comment:
What if the user decides to solve more tasks than they have CPU cores? Do we want to somehow prevent it or at least warn about it?

Reply:
That's fine, and one of the reasons of using SolverManager is to be able to solve more problems than cpu cores OOTB.
In this case there are two policies:
- FIFO: solve as many problems as the number of available cores completely before starting queued up problems (current approach by setting the `FixedThreadPool` size in the `solverExectorService` to #cores).
- Round Robin: Solve all submitted tasks in parallel and let the JVM handle scheduling threads on available cores, this however, doesn't guarantee that each problem will get actual cpu time as configured in the Termination policy.